### ISSUE TITLE
TreeBehavior calculates wrong values for lft/rght when called concurrently

### ISSUE THREAD

**Issue by nook24** [cakephp/cakephp#17174]
### Description

Hi,

i have noticed a problematic behavior (bug?) within the TreeBehavior when concurrent requests are used. In my case, CakePHP is used to provide the backend API to the frontend, and the frontend can send multiple concurrent requests to the same API endpoint.

This leads to all sort of random issues like having multiple nodes with the same values for `lft` and `rght`, or nodes where `lft` is greater-than than `rght`.
I was only able to reproduce this while deleting nodes. I'm not sure why to be honest. This should also happen when creating new nodes.

![broke_tree](https://github.com/cakephp/cakephp/assets/9019992/e624edca-dbb7-4aa0-a6f4-49f0be985a55)

To prevent this issue, some sort of locking is required. I decided to go with a Semaphor [sem_get](https://www.php.net/manual/en/function.sem-get.php).

My plan was to implement the locking via the `beforeDelete` and `afterDelete` callbacks of the table, where I noticed the next bug.
It looks like, that the TreeBehavior is not calling the `beforeSave`and `afterSave` events correctly when the `TreeBehavior` has to updated existing nodes.

So, I still have the same issue, when I implement the locking like so:
```php
<?php

namespace App\Model\Table;

class ContainersTable extends Table {

    public function initialize(array $config): void {
        parent::initialize($config);

        $this->setTable('containers');
        $this->setDisplayField('name');
        $this->setPrimaryKey('id');

        $this->addBehavior('Tree', [
            'cascadeCallbacks' => true // I tried true and false - same results
        ]);

    }

    public function beforeSave(EventInterface $event, EntityInterface $entity, \ArrayObject $options) {
        Log::debug('beforeSave - Acquire table lock');
        $this->acquireLock();
    }

    public function afterSave(EventInterface $event, EntityInterface $entity, \ArrayObject $options) {
        Log::debug(' afterSave - Release table lock');
        $this->releaseLock();
    }

    public function beforeDelete(EventInterface $event, EntityInterface $entity, \ArrayObject $options) {
        Log::debug('beforeDelete - Acquire table lock');
        $this->acquireLock();
    }

    public function afterDelete(EventInterface $event, EntityInterface $entity, \ArrayObject $options) {
        Log::debug('afterDelete - Release table lock');
        $this->releaseLock();
    }
    

    /**
     * Will create a Semaphor with a limit of 1 to avoid multiple operations running concurrently on the containers table
     * This method blocks (if necessary) until the semaphore can be acquired
     *
     * https://github.com/cakephp/cakephp/issues/14983#issuecomment-1613491168
     * @return bool
     */
    public function acquireLock() {
        if ($this->semaphoreKey === null) {
            $this->semaphoreKey = ftok(__FILE__, 'c');
            $this->semaphore = sem_get($this->semaphoreKey, 1, 0666, true);
        }

        return sem_acquire($this->semaphore);
    }

    /**
     * @return bool
     */
    public function releaseLock() {
        if (!is_null($this->semaphore)) {
            return sem_release($this->semaphore);
        }
    }
}

```

The locking works as expected, when im implement it inside the Controller:
```php
<?php

namespace App\Controller;

class ServicetemplategroupsController extends AppController {

    public function delete($id = null) {

        /** @var ContainersTable $ContainersTable */
        $ContainersTable = TableRegistry::getTableLocator()->get('Containers');

        $ContainersTable->acquireLock();

        /** @var $ContainersTable ContainersTable */
        $ContainersTable = TableRegistry::getTableLocator()->get('Containers');
        $container = $ContainersTable->get($id, [
            'contain' => [
                'Servicetemplategroups'
            ]
        ]);

        $ContainersTable->delete($container);

        $ContainersTable->releaseLock();
    }
}
```

But this means I have to dig through all the code to find all parts that are modifying the tree. 

**Edit**: There was also an issue in the past with the same problem: https://github.com/cakephp/cakephp/issues/14983

**Update**
Looks like the tree behavior is calling the `_sync` method, before the `beforeDelete` method in the Table gets executed

![tree_sync](https://github.com/cakephp/cakephp/assets/9019992/26d4d6c2-edfc-4f0c-9976-33c944caf5bb)

### CakePHP Version

4.4.14

### PHP Version

8.1

**Comment by LordSimal** [comment]
Sounds to me like a race condition problem since this only happens with multiple php processes, right?

**Comment by nook24** [comment]
Yes, it totally is. The Tree Behavior should contain a warning that it is not thread safe / safe for concurrent requests.
I mean, most web applications are using Apache Mod PHP or php-fpm. Both have multiple php worker processes and can therefore crash the Tree Behavior.

Also the TreeBehavior should call the events from the Table whenever it manipulate records, which it not does in all cases.

**Comment by markstory** [comment]
> The Tree Behavior should contain a warning that it is not thread safe / safe for concurrent requests.

Sure we can update the documentation, I don't think there is a good way for the framework to implement locking which will leave us needing to document how to perform locking.

**Comment by othercorey** [comment]
Could we solve this with transactions somehow?

**Comment by nook24** [comment]
I was also thinking about transactions. I guess this would require to do everything based on SQL. Someone suggested this for Laravel: https://github.com/lazychaser/laravel-nestedset/issues/503

**Comment by markstory** [comment]
> Could we solve this with transactions somehow?

Not in most database deployments. Solving this with transactions requires `serializable` transaction isolation. Repeatable read, and read committed would still allow concurrent writes that could result in a corrupt tree. Most databases are not run in serializable mode because it creates a huge performance hit.

From your link @nook24 

>  perform all calculations on the database side with row locking during update.

The row locking for update `SELECT FOR UPDATE` style query will solve the concurrency issue here. We don't have an abstraction for row locking but could be a useful improvement if it helps put this issue to rest.

**Comment by LordSimal** [comment]
I found this article today related to `SELECT FOR UPDATE` in postgres
https://www.cybertec-postgresql.com/en/select-for-update-considered-harmful-postgresql

**Comment by dereuromark** [comment]
anything to be done here?

**Comment by nook24** [comment]
At least we should add a warning to the documentation about this. I will send over a PR.

**Comment by LordSimal** [comment]
Thanks @nook24 for the docs PR. That's really all we can do in my opinion.
